### PR TITLE
ENH: WSI Conversion to tiff files

### PR DESCRIPTION
**Pull request by kenza-bouzid**
- Add ConvertWSIToTiffd transform 
- Add TiffConversionConfig for running on a slides dataset
- unit tests

### CODE REVIEW

File: hi-ml-cpath/src/health_cpath/preprocessing/tiff_conversion.py [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+        add_lowest_magnification: bool = False,
+        base_objective_power: Optional[float] = None,
+        replace_ampersand_by: str = UNDERSCORE,
+        compression: COMPRESSION = COMPRESSION.ADOBE_DEFLATE,
+        tile_size: int = 512,
+        backend: WSIBackend = WSIBackend.OPENSLIDE,
+    ) -> None:
+        """
+        :param output_folder: The directory where the tiff file will be saved.
+        :param image_key: The key of the image in the data dictionary, defaults to SlideKey.IMAGE
```
Comment (ant0nsc) [inline hi-ml-cpath/src/health_cpath/preprocessing/tiff_conversion.py:56]:
```suggestion
        :param image_key: The key of the image that should be converted in the data dictionary, defaults to SlideKey.IMAGE
```

File: hi-ml-cpath/src/health_cpath/preprocessing/tiff_conversion.py [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+        :param src_format: The format of the src file, defaults to WSIFormat.NDPI
+        :param target_magnifications: The target magnifications e.g. [10., 20.], defaults to [10.]. If
+        target_magnifications is None, the tiff file will contain the image data at all magnifications.
```
Comment (ant0nsc) [inline hi-ml-cpath/src/health_cpath/preprocessing/tiff_conversion.py:59]: Here and across the whole PR: when doing documentation `:param` that span multiple lines, the second line needs to be indented.

File: hi-ml-cpath/src/health_cpath/preprocessing/tiff_conversion.py [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+        replace_ampersand_by: str = UNDERSCORE,
+        compression: COMPRESSION = COMPRESSION.ADOBE_DEFLATE,
+        tile_size: int = 512,
+        backend: WSIBackend = WSIBackend.OPENSLIDE,
+    ) -> None:
+        """
+        :param output_folder: The directory where the tiff file will be saved.
+        :param image_key: The key of the image in the data dictionary, defaults to SlideKey.IMAGE
+        :param src_format: The format of the src file, defaults to WSIFormat.NDPI
+        :param target_magnifications: The target magnifications e.g. [10., 20.], defaults to [10.]. If
```
Comment (ant0nsc) [inline hi-ml-cpath/src/health_cpath/preprocessing/tiff_conversion.py:58]:
```suggestion
        :param target_magnifications: The magnifications that should be read, converted and written to the output file, e.g. [10., 20.], defaults to [10.]. If
```

File: hi-ml-cpath/src/health_cpath/preprocessing/tiff_conversion.py [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+        base_objective_power: Optional[float] = None,
+        replace_ampersand_by: str = UNDERSCORE,
+        compression: COMPRESSION = COMPRESSION.ADOBE_DEFLATE,
+        tile_size: int = 512,
+        backend: WSIBackend = WSIBackend.OPENSLIDE,
+    ) -> None:
+        """
+        :param output_folder: The directory where the tiff file will be saved.
+        :param image_key: The key of the image in the data dictionary, defaults to SlideKey.IMAGE
+        :param src_format: The format of the src file, defaults to WSIFormat.NDPI
```
[the remark this answers is not in the document]
Comment (kenza-bouzid) [reply hi-ml-cpath/src/health_cpath/preprocessing/tiff_conversion.py:57]: Good point, I'll make it dynamic instead, I can just change the file's extension. thanks

File: hi-ml-cpath/src/health_cpath/preprocessing/tiff_conversion.py [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+        :param target_magnifications: The target magnifications e.g. [10., 20.], defaults to [10.]. If
+        target_magnifications is None, the tiff file will contain the image data at all magnifications.
+        :param add_lowest_magnification: A flag indicating whether the tiff file should also contain the image data at
+        the lowest magnification, defaults to False. This is useful if the lowest magnification of the wsi is not part
+        of the target magnifications and one wants to use the lowest magnification for faster processing.
+        :param base_objective_power: The base objective power of the wsi. This is used to calculate the magnification
```
Comment (ant0nsc) [inline hi-ml-cpath/src/health_cpath/preprocessing/tiff_conversion.py:63]:
```suggestion
        :param default_base_objective_power: The base objective power of the wsi. This is used to calculate the magnification
```

File: hi-ml-cpath/src/health_cpath/preprocessing/tiff_conversion.py [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+        of the wsi. If the objective power is not found in the wsi properties, the base_objective_power is used
+        instead. If the objective power is not found in the wsi properties and base_objective_power is None, an error
+        is raised., defaults to None
+        :param replace_ampersand_by: A string that is used to replace ampersands in the src file name, defaults to
+        UNDERSCORE. This is useful because ampersands in file names can cause problems in cloud storage.
+        :param compression: The compression that is used to save the tiff file, defaults to COMPRESSION.ADOBE_DEFLATE
+        aka ZLIB that is lossless compression.
+        :param tile_size: The size of the tiles that are used to write the tiff file, defaults to 512.
+        :param backend: The backend that is used to read the wsi file, defaults to WSIBackend. OPENSLIDE which is
```
Comment (ant0nsc) [inline hi-ml-cpath/src/health_cpath/preprocessing/tiff_conversion.py:72]:
```suggestion
        :param backend: The backend that is used to read the wsi file, defaults to `WSIBackend.OPENSLIDE` which is
```

File: hi-ml-cpath/src/health_cpath/preprocessing/tiff_conversion.py [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+    def _get_base_objective_power(self, wsi_obj: OpenSlide) -> float:
+        """Returns the objective power of the wsi. The objective power is extracted from the wsi properties. If the
+        objective power is not found in the wsi properties, the base_objective_power is used instead.
+
+        :param wsi_obj: The wsi object in openslide format
+        :raises ValueError: Raises an error if the objective power is not found in the wsi properties and
+        base_objective_power is None
+        :return: The base objective power of the wsi
+        """
+        base_objective_power = wsi_obj.properties.get(self.OBJECTIVE_POWER_KEY, self.base_objective_power)
```
Comment (ant0nsc) [inline hi-ml-cpath/src/health_cpath/preprocessing/tiff_conversion.py:109]: I'm confused. YOu're reading that out with a key that is `openslide....` does that exist if the backend is `cucim`?

File: hi-ml-cpath/src/health_cpath/preprocessing/tiff_conversion.py [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+        and will be written as tags in the tiff file.
+
+        :param wsi_obj: The wsi object in openslide format
+        :raises ValueError: Raises an error if the resolution unit is not in centimeters
+        :return: A dictionary of options that will be passed to the tiffwriter
+        """
+        resolution_unit = wsi_obj.properties[self.RESOLUTION_UNIT_KEY]
+
+        if resolution_unit != self.RESOLUTION_UNIT:
+            raise ValueError(f"Resolution unit is not in {self.RESOLUTION_UNIT}: {resolution_unit}")
```
Comment (ant0nsc) [inline hi-ml-cpath/src/health_cpath/preprocessing/tiff_conversion.py:185]: just curious, why is this check necessary? Couldn't we just pass through the unit from input to output?

File: hi-ml-cpath/src/health_cpath/preprocessing/tiff_conversion.py [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+        return options
+
+    def get_px_per_cm_resolution_at_level(self, wsi_obj: OpenSlide, level: int) -> Tuple[float, float]:
+        """Returns the resolution of the wsi at a given level in pixels per centimeter.
+
+        :param wsi_obj: The wsi object in openslide format
+        :param level: The level at which the resolution is calculated
+        :return: A tuple of floats (x_resolution, y_resolution)
+        """
+        um_per_cm = 10000
```
Comment (ant0nsc) [inline hi-ml-cpath/src/health_cpath/preprocessing/tiff_conversion.py:204]: maybe add a check here the resolution unit is really centimetres?

File: hi-ml-cpath/src/health_cpath/preprocessing/tiff_conversion.py [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+        """Makes sure that the level data is in HWC format. If the level data is not in HWC format, an error is raised.
+        This sanity check is necessary to avoid compression issues.
+        """
+        if level_data.shape[2] != 3:
+            raise ValueError(
+                f"Expected 3 channels but got {level_data.shape[2]}. Maybe the image is in channel first format? Try "
+                "transposing the image."
+            )
+
+    def get_options(self, wsi_obj: OpenSlide) -> Dict[str, Any]:
```
Comment (ant0nsc) [inline hi-ml-cpath/src/health_cpath/preprocessing/tiff_conversion.py:174]:
```suggestion
    def get_tiffwriter_options(self, wsi_obj: OpenSlide) -> Dict[str, Any]:
```

File: hi-ml-cpath/src/health_cpath/utils/tiff_conversion_config.py [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+        default=COMPRESSION.ADOBE_DEFLATE,
+        class_=COMPRESSION,
+        doc="The compression that will be used for the tiff files. Default is ADOBE_DEFLATE for lossless compression.",
+    )
+    tile_size: int = param.Integer(
+        default=512, doc="The size of the tiles used to write the tiff file, defaults to 512."
+    )
+    num_workers: int = param.Integer(
+        default=1,
+        doc="The number of workers that will be used to convert the src files to tiff files. If num_workers is 1.",
```
Comment (ant0nsc) [inline hi-ml-cpath/src/health_cpath/utils/tiff_conversion_config.py:59]: incomplete sentence?

File: hi-ml-cpath/src/health_cpath/utils/tiff_conversion_config.py [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+
+
+class TiffConversionConfig(param.Parameterized):
+    image_key: SlideKey = param.ClassSelector(
+        class_=SlideKey,
+        default=SlideKey.IMAGE,
+        doc="The key of the image in the dataset. This is used to get the path of the src file."
+    )
+    src_format: WSIFormat = param.ClassSelector(
+        class_=WSIFormat, default=WSIFormat.NDPI, doc="The format of the source files. Default is NDPI."
```
Comment (ant0nsc) [inline hi-ml-cpath/src/health_cpath/utils/tiff_conversion_config.py:27]: See my comment in tiff_conversion - this argument name is somewhat misleading, it is only the expected file extension?

File: hi-ml-cpath/src/health_cpath/preprocessing/tiff_conversion.py [hunk 13]
```diff
@@ -0,0 +1,241 @@
...
+
+    def get_tiff_path(self, src_path: Path) -> Path:
+        """Returns the path to the tiff file that will be created from the src file. The tiff file is saved in the
+        output_folder with the same name as the src file but with the tiff extension. Ampersands are replaced by the
+        replace_ampersand_by string.
+
+        :param src_path: The path to the src file.
+        :return: The path to the tiff file that will be created from the src file.
+        """
+        tiff_filename = src_path.name.replace(self.src_format, WSIFormat.TIFF)
```
Comment (ant0nsc) [inline hi-ml-cpath/src/health_cpath/preprocessing/tiff_conversion.py:96]: You only want to change the extension/suffix, so it is safer to use `Path.with_suffix` https://docs.python.org/3/library/pathlib.html#pathlib.PurePath.with_suffix

File: hi-ml-cpath/testhisto/testhisto/preprocessing/test_tiff_conversion.py [hunk 14]
```diff
@@ -0,0 +1,215 @@
...
+
+
+def test_base_objective_power(wsi_samples: WSISamplesType) -> None:
+    target_mag = 2.5
+    transform = ConvertWSIToTiffd(
+        output_folder=Path("foo"),
+        image_key=SlideKey.IMAGE,
+        src_format=WSIFormat.TIFF,
+        target_magnifications=[target_mag],
+        base_objective_power=None,
```
Comment (ant0nsc) [inline hi-ml-cpath/testhisto/testhisto/preprocessing/test_tiff_conversion.py:56]: Here and in other tests: you are providing arguments to ConvertWSIToTiffd that are already the default values. The tests
would be easier to read if you only provide those that matter to your test. even output_folder could have a default
value that is "." to further simplify.

**Review by ant0nsc — APPROVED** [review]
Looking great! just a few comments on corner cases and docs.

File: hi-ml-cpath/src/health_cpath/preprocessing/tiff_conversion.py (hunk 7 above)
Comment (kenza-bouzid) [reply hi-ml-cpath/src/health_cpath/preprocessing/tiff_conversion.py:109]: That is why we have default_base_objectiove_power to use if that field is not part of the properties. 
I haven't been able to find the right tag to set for base objective power for tiff files, https://www.loc.gov/preservation/digital/formats/content/tiff_tags.shtml#:~:text=TIFF%20files%20declare%20and%20describe%20their%20content%20by,or%20another%20image%20compression%20option%20has%20been%20used.
Here's the available metadata using cucim 
image
This OBJECTIVE_POWER_KEY is only available in the original ndpi files. 

Now that you said that I haven't tested the cpode with a different backend and I think it won't work given that all wsi_obj are typed as OpenSlide

I just made this change (making backend flexible) but I think we better stick with Openslide

File: hi-ml-cpath/src/health_cpath/preprocessing/tiff_conversion.py (hunk 9 above)
Comment (kenza-bouzid) [reply hi-ml-cpath/src/health_cpath/preprocessing/tiff_conversion.py:204]: We check that in get_options from the properties

File: hi-ml-cpath/src/health_cpath/utils/tiff_conversion_config.py (hunk 11 above)
Comment (kenza-bouzid) [reply hi-ml-cpath/src/health_cpath/utils/tiff_conversion_config.py:59]: fixed thanks!

File: hi-ml-cpath/src/health_cpath/utils/tiff_conversion_config.py (hunk 12 above)
Comment (kenza-bouzid) [reply hi-ml-cpath/src/health_cpath/utils/tiff_conversion_config.py:27]: removed this param!

File: hi-ml-cpath/src/health_cpath/preprocessing/tiff_conversion.py (hunk 13 above)
Comment (kenza-bouzid) [reply hi-ml-cpath/src/health_cpath/preprocessing/tiff_conversion.py:96]: oh didn't know about this, it's a lot nice, thanks!

File: hi-ml-cpath/testhisto/testhisto/preprocessing/test_tiff_conversion.py (hunk 14 above)
Comment (kenza-bouzid) [reply hi-ml-cpath/testhisto/testhisto/preprocessing/test_tiff_conversion.py:56]: I am a bit too cautious with default values, if the default values change for some reason the test might break! It should be fairly easy to fix but it took me one hour once to figure out why my test didn't pass (it was because of the default values changing). I usually use default values that are useful to run the script with minimum command line arguments (hence the tile size deault to 512 that worked best for ndpi conversion) but try to reset these in the tests, but agreed I can simplify tests that only check exceptions for example.

File: hi-ml-cpath/src/health_cpath/preprocessing/tiff_conversion.py (hunk 7 above)
Comment (ant0nsc) [reply hi-ml-cpath/src/health_cpath/preprocessing/tiff_conversion.py:109]: Agree, let's stick with openslide only for now.

File: hi-ml-cpath/src/health_cpath/preprocessing/tiff_conversion.py [hunk 15]
```diff
@@ -0,0 +1,241 @@
+from enum import Enum
```
Comment (harshita-s) [inline hi-ml-cpath/src/health_cpath/preprocessing/tiff_conversion.py]: needs Microsoft header

File: hi-ml-cpath/src/health_cpath/preprocessing/tiff_conversion.py [hunk 16]
```diff
@@ -0,0 +1,241 @@
...
+        self.output_folder = output_folder
+        self.image_key = image_key
+        self.src_format = src_format
+        if target_magnifications is not None:
+            target_magnifications.sort(reverse=True)
+        self.target_magnifications = target_magnifications if target_magnifications else None
+        self.add_lowest_magnification = add_lowest_magnification
+        self.replace_ampersand_by = replace_ampersand_by
+        self.base_objective_power = base_objective_power
+        self.wsi_reader = WSIReader(backendbackend=backend)
```
Comment (harshita-s) [inline hi-ml-cpath/src/health_cpath/preprocessing/tiff_conversion.py]: why this is `backendbackend`?

File: hi-ml-cpath/src/health_cpath/preprocessing/tiff_conversion.py [hunk 17]
```diff
@@ -0,0 +1,241 @@
...
+class WSIFormat(str, Enum):
+    """The format of the wsi file."""
+    NDPI = "ndpi"
+    SVS = "svs"
+    TIF = "tif"
+    TIFF = "tiff"
+
+
+class ConvertWSIToTiffd(MapTransform):
+    """Converts a wsi file to a tiff file. The tiff file is saved in the output_folder with the same name as the src
```
Comment (harshita-s) [inline hi-ml-cpath/src/health_cpath/preprocessing/tiff_conversion.py:21]: could you please add supported formats in the docstring

**Review by harshita-s — APPROVED** [review]

File: hi-ml-cpath/src/health_cpath/preprocessing/tiff_conversion.py (hunk 16 above)
Comment (kenza-bouzid) [reply hi-ml-cpath/src/health_cpath/preprocessing/tiff_conversion.py]: typo, fixed!

File: hi-ml-cpath/src/health_cpath/preprocessing/tiff_conversion.py (hunk 17 above)
Comment (kenza-bouzid) [reply hi-ml-cpath/src/health_cpath/preprocessing/tiff_conversion.py:21]: Sure, this should work with all supported format by openslide
